Initial support for Tree Borrows (without protector) (#213)
minirust/minirust: spec/mem/tree_borrows/automata.md

            Permission::Frozen => throw_ub!("Tree Borrows: Child writing a pointer with the Frozen permission"),
        }
    }

    fn foreign_read(permission: Permission) -> Result<Option<Permission>> {
        ret(
            match permission {
                Permission::Reserved => Some(Permission::Reserved),
                Permission::Active => Some(Permission::Frozen),
                Permission::Frozen => Some(Permission::Frozen),
            }
        )
    }

    fn permission_transition(
        &mut self,
        tag: BorTag,
        addr: Address,
        access_kind: AccessKind,
        node_relation: NodeRelation,
    ) -> Result {
        let Some(mut node) = self.tree.nodes.get(tag) else { throw_ub!("Tree Borrows: node not existed"); };

        let Some(curr_permission) = node.permissions.get(addr) else {
            if node_relation == NodeRelation::Foreign {
                return ret(());
            }
            match access_kind {
                AccessKind::Read => throw_ub!("Tree Borrows: Child reading a pointer without permission"),
                AccessKind::Write => throw_ub!("Tree Borrows: Child writing a pointer without permission"),

>             }

Reviewer: Hm, I can see how having this here isn't that great.

What if you change the argument type for `foreign_read` etc to `Option<Permission>` as well? Then this can be handled as before.

        };

        let next_permission = match (node_relation, access_kind) {

>             (NodeRelation::Foreign, AccessKind::Write) => None,

Reviewer: Please still call a `Self::foreign_write` function here, even if it always returns `None` for now.

# State Machine For the Tree Borrows

> The core of the Tree Borrows is a state machine. We track a state machine for each node on the tree. We call the state as *Permission*.

Reviewer: ```suggestion
The core of Tree Borrows is a state machine for each node and each location.
We call the state a *Permission*.
```
